jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/muc/ChatRoomImpl.kt

```diff
@@ -294,35 +294,58 @@ class ChatRoomImpl(
         presenceToSend?.let { xmppProvider.xmppConnection.tryToSendStanza(it) }
     }
 
-    override val presenceExtensions
-        get() = synchronized(this) { lastPresenceSent?.extensions?.toList() ?: emptyList() }
-
-    override fun modifyPresenceExtensions(
-        toRemove: Collection<ExtensionElement>,
-        toAdd: Collection<ExtensionElement>
-    ) {
-        val presenceToSend: Presence? = synchronized(this) {
+    override fun addPresenceExtensionIfMissing(extension: ExtensionElement) {
+        val presenceToSend = synchronized(this) {
             lastPresenceSent?.let { presence ->
-                var changed = false
-                toRemove.forEach {
-                    presence.removeExtension(it)
-                    // We don't have a good way to check if it was actually removed.
-                    changed = true
-                }
-                toAdd.forEach {
-                    presence.addExtension(it)
-                    changed = true
+                if (presence.extensions?.any { it.qName == extension.qName } == true) {
+                    null
+                } else {
+                    presence.addExtension(extension).build()
                 }
-                if (changed) presence.build() else null
-            } ?: run {
-                logger.error("No presence packet obtained yet")
-                null
             }
         }
+        presenceToSend?.let { xmppProvider.xmppConnection.tryToSendStanza(it) }
+    }
+
+    override fun removePresenceExtensions(pred: (ExtensionElement) -> Boolean) {
+        val presenceToSend = synchronized(this) {
+            lastPresenceSent?.extensions?.filter { pred(it) }?.let {
+                modifyPresenceExtensions(toRemove = it)
```

Comment:
This doesn't trigger a ConcurrentModificationException?

Reply:
Not unless I'm misunderstanding something. I accesses `lastPresenceSent?.extensions`, filters the result to a list, and then calls `modifyPresenceExtensions`.

Reply:
Oh, ok, because filter constructs a whole new list object, rather than being another iterator - right.